api/grpcserver/transaction_service.go

```diff
@@ -226,42 +226,46 @@ func (s TransactionService) TransactionsStateStream(in *pb.TransactionsStateStre
 				return nil
 			}
 			// Filter for any matching transactions in the reported layer
-			// TODO: this is inefficient and could be optimized!
-			// See https://github.com/spacemeshos/go-spacemesh/issues/2076
 			for _, b := range layer.Layer.Blocks() {
-				for _, layerTxid := range b.TxIDs {
-					for _, txid := range in.TransactionId {
-						if bytes.Equal(layerTxid.Bytes(), txid.Id) {
-							var txstate pb.TransactionState_TransactionState
-							switch layer.Status {
-							case events.LayerStatusTypeApproved:
-								txstate = pb.TransactionState_TRANSACTION_STATE_MESH
-							case events.LayerStatusTypeConfirmed:
-								txstate = pb.TransactionState_TRANSACTION_STATE_PROCESSED
-							default:
-								txstate = pb.TransactionState_TRANSACTION_STATE_UNSPECIFIED
-							}
-							res := &pb.TransactionsStateStreamResponse{
-								TransactionState: &pb.TransactionState{
-									Id:    txid,
-									State: txstate,
-								},
-							}
-							if in.IncludeTransactions {
-								tx, err := s.Mesh.GetTransaction(layerTxid)
-								if err != nil {
-									log.Error("could not find transaction %v from layer %v: %v", layerTxid, layer, err)
-									return status.Error(codes.Internal, "error retrieving tx data")
-								}
+				blockTXIDSet := make(map[types.TransactionID]struct{})
 
-								res.Transaction = convertTransaction(tx)
-							}
-							if err := stream.Send(res); err != nil {
-								return err
+				//create a set for the block transaction IDs
+				for _, txid := range b.TxIDs {
+					blockTXIDSet[txid] = struct{}{}
+				}
+
+				for _, inputTxID := range in.TransactionId {
+					var arrayID [32]byte
+					copy(arrayID[:], inputTxID.Id[:])
+					txid := types.TransactionID(arrayID)
+					// if there is an ID corresponding to inputTxID in the block
+					if _, exists := blockTXIDSet[txid]; exists {
+						var txstate pb.TransactionState_TransactionState
+						switch layer.Status {
```

Comment:
Another potential optimization here: this `switch` statement only depends on the `layer.Status`. It doesn't need to run for every tx. Could you move this outside the loop?